x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/queries/SparseVectorQueryBuilder.java

```diff
@@ -0,0 +1,369 @@
+                            listener.onResponse(null);
+                        } else if (inferenceResponse.getInferenceResults().get(0) instanceof WarningInferenceResults warning) {
+                            listener.onFailure(new IllegalStateException(warning.getWarning()));
+                        } else {
+                            listener.onFailure(
+                                new IllegalStateException(
+                                    "expected a result of type ["
+                                        + TextExpansionResults.NAME
+                                        + "] received ["
+                                        + inferenceResponse.getInferenceResults().get(0).getWriteableName()
+                                        + "]. Is ["
+                                        + modelId
+                                        + "] a compatible model?"
+                                )
+                            );
+                        }
+                    }, listener::onFailure)
+                )
+            );
+
+            return new SparseVectorQueryBuilder(this, textExpansionResultsSupplier);
+
+        } else {
+            return vectorDimensionsToQuery(fieldName, vectorDimensions);
+        }
+
+    }
+
+    private QueryBuilder vectorDimensionsToQuery(String fieldName, List<VectorDimension> vectorDimensions) {
+        if (tokenPruningConfig != null) {
+            VectorDimensionsQueryBuilder vectorDimensionsQueryBuilder = new VectorDimensionsQueryBuilder(
+                fieldName,
+                vectorDimensions,
+                tokenPruningConfig
+            );
+            vectorDimensionsQueryBuilder.queryName(queryName);
+            vectorDimensionsQueryBuilder.boost(boost);
+            return vectorDimensionsQueryBuilder;
+        }
+        // Note: Weighted tokens queries were introduced in 8.13.0. To support mixed version clusters prior to 8.13.0,
+        // if no token pruning configuration is specified we fall back to a boolean query.
+        // TODO this should be updated to always use a WeightedTokensQueryBuilder once it's in all supported versions.
+        var boolQuery = QueryBuilders.boolQuery();
+        for (var weightedToken : vectorDimensions) {
+            boolQuery.should(QueryBuilders.termQuery(fieldName, weightedToken.token()).boost(weightedToken.weight()));
+        }
+        boolQuery.minimumShouldMatch(1);
+        boolQuery.boost(boost);
+        boolQuery.queryName(queryName);
+        return boolQuery;
+    }
+
+    private QueryBuilder textExpansionResultsToQuery(String fieldName, TextExpansionResults textExpansionResults) {
+        return vectorDimensionsToQuery(fieldName, textExpansionResults.getVectorDimensions());
+    }
+
+    @Override
+    protected Query doToQuery(SearchExecutionContext context) {
+        throw new IllegalStateException("sparse_vector should have been rewritten to another query type");
+    }
```

Comment:
I don't think we should rewrite this to another query type. Instead we should serialize down the query vector and create the appropriate query.

Reply:
Are there any BwC management benefits to rewriting to another query type?

Reply:
I see your point here. I also see it as a little of a chicken and egg scenario, since `WeightedTokensQueryBuilder` will be refactored after this PR is merged, and the logic will be shared with both. I didn't want to add that refactoring to this PR as it's already quite large.